controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -240,18 +254,32 @@ func (ew *EndpointsWatcher) deleteService(obj interface{}) {
 }
 
 func (ew *EndpointsWatcher) addEndpoints(obj interface{}) {
-	endpoints := obj.(*corev1.Endpoints)
-	if endpoints.Namespace == kubeSystem {
-		return
-	}
-	id := ServiceID{
-		Namespace: endpoints.Namespace,
-		Name:      endpoints.Name,
-	}
+	var sp *servicePublisher
+	switch res := obj.(type) {
+	case *corev1.Endpoints:
+		{
+			if res.Namespace == kubeSystem {
+				return
+			}
+			id := ServiceID{res.Namespace, res.Name}
+			sp = ew.getOrNewServicePublisher(id)
+		}
 
-	sp := ew.getOrNewServicePublisher(id)
+	case *discovery.EndpointSlice:
+		{
+			if res.Namespace == kubeSystem {
+				return
+			}
+			id, err := getEndpointSliceServiceID(res)
+			if err != nil {
+				ew.log.Errorf("Could not fetch service name for EndpointSlice [%s/%s]:%v", res.Namespace, res.Name, err)
+				return
+			}
+			sp = ew.getOrNewServicePublisher(id)
+		}
+	}
 
-	sp.updateEndpoints(endpoints)
+	sp.updateEndpoints(obj)
```

Comment:
Because Golang is amazing, you can pass some arbitrary interface here in this function that is NOT `discovery.EndpointSlice` or `corev1.Endpoints` and skip on both of the match arms, which would leave your sp uninitialized and you will git an NPE. I think you want a default case here that handles that  somehow.

Reply:
Yup, you're completely right. Will change this asap.

I thought we would always get either a `discovery.EndpointSlice` or `corev1.Endpoints` since the function is registered to the handler, but perhaps it's better not to take any chances. I'll edit it so that each match arm has a separate call to updateEndpoints.